runtime/compiler/trj9/optimizer/JProfilingValue.cpp

```diff
@@ -264,34 +264,56 @@ void
 TR_JProfilingValue::lowerCalls()
    {
    TR::TreeTop *cursor = comp()->getStartTree();
+   int32_t numberOfNodes = comp()->getFlowGraph()->getNextNodeNumber();
+   TR_BitVector *backwardAnalyzedListOfAddressNodes = new (comp()->trStackMemory())TR_BitVector();
    while (cursor)
       {
       TR::Node * node = cursor->getNode();
-
       if (node->isProfilingCode() &&
           node->getOpCodeValue() == TR::treetop &&
           node->getFirstChild()->getOpCode().isCall() &&
           (node->getFirstChild()->getSymbolReference()->getReferenceNumber() == TR_jProfile32BitValue ||
            node->getFirstChild()->getSymbolReference()->getReferenceNumber() == TR_jProfile64BitValue) &&
           node->getFirstChild()->getNumChildren() == 3)
          {
-         TR::TreeTop *nextTree = cursor->getNextTreeTop();
+         // Backward Analysis in extended BBs to get list of address nodes 
+         for (TR::TreeTop *iter = cursor->getPrevTreeTop();
+            iter && (iter->getNode()->getOpCodeValue() != TR::BBStart || iter->getNode()->getBlock()->isExtensionOfPreviousBlock());
+            iter=iter->getPrevTreeTop())
+            {
+            TR::Node *currentTreeTopNode = iter->getNode();
+            if (currentTreeTopNode->getNumChildren() >= 1 && currentTreeTopNode->getFirstChild()->getType() == TR::Address)
+               backwardAnalyzedListOfAddressNodes->set(currentTreeTopNode->getFirstChild()->getGlobalIndex());
+            }
+         // Forward Analysis in extended BBs to get a list of compressedref anchor nodes 
+         for (TR::TreeTop *iter = cursor->getNextTreeTop();
+            iter && (iter->getNode()->getOpCodeValue() != TR::BBStart || iter->getNode()->getBlock()->isExtensionOfPreviousBlock());
+            iter=iter->getNextTreeTop())
+            {
+            TR::Node *currentTreeTopNode = iter->getNode();
+            if (currentTreeTopNode->getOpCodeValue() == TR::compressedRefs &&
+               backwardAnalyzedListOfAddressNodes->isSet(currentTreeTopNode->getFirstChild()->getGlobalIndex()))
```

Comment:
it might be easier to read this if the && is moved to the start of the isSet check like is done for many long conditions in the optimizer